### PR TITLE
Set Sec-Websocket-Protocol header according RFC

### PR DESCRIPTION
**Pull request by arcusfelis**
This PR addresses #1221 

Proposed changes include:
* This PR fixes compatibility issues with some Websocket libraries (chrome, some dot net stuff).
* Do not set `Sec-Websocket-Protocol` header in response if it's missing in request.
* Set `Sec-Websocket-Protocol` header with xmpp value, if it is specified in request.
* Do not set `Sec-Websocket-Protocol` header in respond if request does not specify xmpp.

Some libraries does not set Sec-Websocket-Protocol, but check that Sec-Websocket-Protocol exists in the result. It confuses the libraries and they disconnect from the server with "cryptic" error description.

Spec:

```
   The |Sec-WebSocket-Protocol| header field is used in the WebSocket
   opening handshake.  It is sent from the client to the server and back
   from the server to the client to confirm the subprotocol of the
   connection.  This enables scripts to both select a subprotocol and be
   sure that the server agreed to serve that subprotocol.

   The |Sec-WebSocket-Protocol| header field MAY appear multiple times
   in an HTTP request (which is logically the same as a single
   |Sec-WebSocket-Protocol| header field that contains all values).
   However, the |Sec-WebSocket-Protocol| header field MUST NOT appear
   more than once in an HTTP response.
```

More interesting part of the spec:

```
       /subprotocol/
          Either a single value representing the subprotocol the server
          is ready to use or null.  The value chosen MUST be derived
          from the client's handshake, specifically by selecting one of
          the values from the |Sec-WebSocket-Protocol| field that the
          server is willing to use for this connection (if any).  If the
          client's handshake did not contain such a header field or if
          the server does not agree to any of the client's requested
          subprotocols, the only acceptable value is null.  The absence
          of such a field is equivalent to the null value (meaning that
          if the server does not wish to agree to one of the suggested
          subprotocols, it MUST NOT send back a |Sec-WebSocket-Protocol|
          header field in its response).  The empty string is not the
          same as the null value for these purposes and is not a legal
          value for this field.  The ABNF for the value of this header
          field is (token), where the definitions of constructs and
          rules are as given in [RFC2616].

```

from here https://tools.ietf.org/html/rfc6455

### CODE REVIEW

File: apps/ejabberd/src/mod_websockets.erl [hunk 1]
```diff
@@ -329,3 +329,35 @@ maybe_send_ping_request(none) ->
...
+                    cowboy_req:set_resp_header(<<"Sec-WebSocket-Protocol">>, <<"xmpp">>, Req1);
+                false ->
+                    %% Do not agree with client, do not add a respond header
+                    ?DEBUG("issue=not_supported_procol, protocol=~p", [Proto]),
+                    Req1
+            end
+    end.
+
+split_protocol(Proto) ->
+	%% whitespace and comma can be used as separators
```
Comment (fenek) [inline apps/ejabberd/src/mod_websockets.erl]: Tab instead of spaces.

**Comment by michalwski** [comment]
I agree with @fenek, this change requires tests.

**Comment by arcusfelis** [comment]
Original PR for cowboy 1:

```erlang
commit 686518f092024d683fb15d0d7c7e3dc640c368a7 (origin/fix-websocket-sec-header, fix-websocket-sec-header)
Author: Michael Uvarov <arcusfelis@gmail.com>
Date:   Thu May 4 18:00:49 2017 +0200

    Set Sec-Websocket-Protocol header according RFC

    This PR fixes compatibility issues with some Websocket libraries.
    Do not set Sec-Websocket-Protocol header in respond if it's missing in request.
    Do not set Sec-Websocket-Protocol header in respond if it's not xmpp in
    request.

diff --git a/apps/ejabberd/src/mod_websockets.erl b/apps/ejabberd/src/mod_websockets.erl
index b0bea1343..1f2770454 100644
--- a/apps/ejabberd/src/mod_websockets.erl
+++ b/apps/ejabberd/src/mod_websockets.erl
@@ -76,7 +76,7 @@ terminate(_Reason, _Req, _State) ->
 % Called for every new websocket connection.
 websocket_init(Transport, Req, Opts) ->
     ?DEBUG("websocket_init: ~p~n", [{Transport, Req, Opts}]),
-    Req1 = cowboy_req:set_resp_header(<<"Sec-WebSocket-Protocol">>, <<"xmpp">>, Req),
+    Req1 = add_sec_websocket_protocol_header(Req),
             Timeout = gen_mod:get_opt(timeout, Opts, infinity),
             PingRate = gen_mod:get_opt(ping_rate, Opts, none),
             ?DEBUG("ping rate is ~p", [PingRate]),
@@ -329,3 +329,35 @@ maybe_send_ping_request(none) ->
     ok;
 maybe_send_ping_request(PingRate) ->
     send_ping_request(PingRate).
+
+
+
+
+
+
+add_sec_websocket_protocol_header(Req) ->
+    case cowboy_req:header(<<"sec-websocket-protocol">>, Req) of
+        {undefined, Req1} ->
+            %% Server should not set "sec-websocket-protocol" respond header
+            %% if it does not present in the request
+            Req1;
+        {Proto, Req1} ->
+            %% A client can propose several protocols
+            %% but we are interested in xmpp only
+            Protocols = split_protocol(Proto),
+            case lists:member(<<"xmpp">>, Protocols) of
+                true ->
+                    cowboy_req:set_resp_header(<<"Sec-WebSocket-Protocol">>, <<"xmpp">>, Req1);
+                false ->
+                    %% Do not agree with client, do not add a respond header
+                    ?DEBUG("issue=not_supported_procol, protocol=~p", [Proto]),
+                    Req1
+            end
+    end.
+
+split_protocol(Proto) ->
+       %% whitespace and comma can be used as separators
+    map_tolower(skip_empty(binary:split(Proto, [<<" ">>, <<",">>], [global]))).
+
+skip_empty(List) ->
+    [X || X <- List, X =/= <<>>].
+
+map_tolower(List) ->
+    [stringprep:tolower(X) || X <- List].
```

**Comment by mongoose-im** [comment]
[5828.1](https://travis-ci.org/esl/MongooseIM/jobs/452110576) / Erlang 19.3 / small_tests / c424d25cb8330f2b91a77c96d4ed4f440adcb239
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1308/5828/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1308/5828/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-07_21.56.10/lib.mongooseim.logs/run.2018-11-07_21.56.11/suite.log.html)

File: src/mod_websockets.erl [hunk 2]
```diff
@@ -358,3 +358,38 @@ maybe_send_ping_request(none) ->
...
+         Proto ->
+             %% A client can propose several protocols
+             %% but we are interested in xmpp only
+             Protocols = split_protocol(Proto),
+             case lists:member(<<"xmpp">>, Protocols) of
+                 true ->
+                     cowboy_req:set_resp_header(<<"Sec-WebSocket-Protocol">>, <<"xmpp">>, Req);
+                 false ->
+                     %% Do not agree with client, do not add a respond header
+                     ?DEBUG("issue=not_supported_procol, protocol=~p", [Proto]),
```
Comment (michalwski) [inline src/mod_websockets.erl]: Is it enough to not set the resp header maybe an error should be send to the client? Is it documented somewhere?

File: src/mod_websockets.erl [hunk 3]
```diff
@@ -358,3 +358,38 @@ maybe_send_ping_request(none) ->
...
+add_sec_websocket_protocol_header(Req) ->
+     case cowboy_req:header(<<"sec-websocket-protocol">>, Req) of
+         undefined ->
+             %% Server should not set "sec-websocket-protocol" respond header
+             %% if it does not present in the request
+             Req;
+         Proto ->
+             %% A client can propose several protocols
+             %% but we are interested in xmpp only
+             Protocols = split_protocol(Proto),
```
Comment (michalwski) [inline src/mod_websockets.erl]: There is `cowboy_req:parse_header/3` function which seems to be doing what you need here. Did you try it @arcusfelis?

File: test/mod_websockets_SUITE.erl [hunk 4]
```diff
@@ -153,3 +181,76 @@ get_ranch_connections() ->
...
+%% From RFC 6455:
+%%   The |Sec-WebSocket-Protocol| header field MAY appear multiple times
+%%   in an HTTP request (which is logically the same as a single
+%%   |Sec-WebSocket-Protocol| header field that contains all values).
+%%   However, the |Sec-WebSocket-Protocol| header field MUST NOT appear
+%%   more than once in an HTTP response.
+agree_to_xmpp_subprotocol_from_many(_) ->
+    check_subprotocol("Two protocols in one header", ["xmpp, other"], "xmpp"),
+    check_subprotocol("Two protocols in one header", ["other, xmpp"], "xmpp"),
+    check_subprotocol("Two protocols in two headers", ["other", "xmpp"], "xmpp").
```
Comment (michalwski) [inline test/mod_websockets_SUITE.erl]: I don't think it's different form the one above. Did I miss sth?

Comment (arcusfelis) [reply test/mod_websockets_SUITE.erl]: First is one header with two subprotocols (i.e. a list with one element)

Second is two headers with one protocol in each (two elements in the list)

Comment (michalwski) [reply test/mod_websockets_SUITE.erl]: Now I see, thanks!

File: src/mod_websockets.erl (hunk 2 above)
Comment (arcusfelis) [reply src/mod_websockets.erl]: Documented. We reply with nil or no header. Client can continue on it’s own risk or disconnect

File: src/mod_websockets.erl (hunk 3 above)
Comment (arcusfelis) [reply src/mod_websockets.erl]: Good catch

**Comment by mongoose-im** [comment]
[5831.1](https://travis-ci.org/esl/MongooseIM/jobs/452372112) / Erlang 19.3 / small_tests / d601749a9d46f9ac1ce8078c40069ea4b53951b1
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1308/5831/small_tests.19.3)

File: src/mod_websockets.erl [hunk 5]
```diff
@@ -365,31 +365,29 @@ maybe_send_ping_request(PingRate) ->
...
- split_protocol(Proto) ->
-     %% whitespace and comma can be used as separators
-    map_tolower(skip_empty(binary:split(Proto, [<<" ">>, <<",">>], [global]))).
-
- skip_empty(List) ->
-     [X || X <- List, X =/= <<>>].
-
- map_tolower(List) ->
-     [stringprep:tolower(X) || X <- List].
+case_insensitive_match(LowerPattern, [Case|Cases]) ->
```
Comment (michalwski) [inline src/mod_websockets.erl]: Spaces around `|` may improve readability.

**Comment by mongoose-im** [comment]
[5832.1](https://travis-ci.org/esl/MongooseIM/jobs/452373619) / Erlang 19.3 / small_tests / 344c935df7591fc435d5d5e4c48a76fd252fd17f
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1308/5832/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1308/5832/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-08_13.12.27/lib.mongooseim.logs/run.2018-11-08_13.12.27/suite.log.html)

**Comment by mongoose-im** [comment]
[5833.1](https://travis-ci.org/esl/MongooseIM/jobs/452389252) / Erlang 19.3 / small_tests / 1b4f60cd81ad678ba4494c167c1fc4cc7549cd17
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1308/5833/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1308/5833/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-08_13.45.03/lib.mongooseim.logs/run.2018-11-08_13.45.04/suite.log.html)

**Pull request merged**